src/dagmc/DagMC.cpp

```diff
@@ -0,0 +1,774 @@
+  rval = MBI->tag_set_by_ptr(tag, &eh, 1, &p, &new_len);
+  delete[] new_packed_string;
+  return rval;
+}
+
+ErrorCode DagMC::unpack_packed_string(Tag tag, EntityHandle eh,
+                                      std::vector< std::string >& values) {
+  ErrorCode rval;
+  const void* p;
+  const char* str;
+  int len;
+  rval = MBI->tag_get_by_ptr(tag, &eh, 1, &p, &len);
+  if (rval != MB_SUCCESS)
+    return rval;
+  str = static_cast<const char*>(p);
+  int idx = 0;
+  while (idx < len) {
+    std::string item(str + idx);
+    values.push_back(item);
+    idx += item.length() + 1;
+  }
+  return MB_SUCCESS;
+}
+
+ErrorCode DagMC::parse_properties(const std::vector<std::string>& keywords,
+                                  const std::map<std::string, std::string>& keyword_synonyms,
+                                  const char* delimiters) {
+  ErrorCode rval;
+
+  // master keyword map, mapping user-set words in cubit to canonical property names
+  std::map< std::string, std::string > keyword_map(keyword_synonyms);
+
+  for (std::vector<std::string>::const_iterator i = keywords.begin();
+       i != keywords.end(); ++i) {
+    keyword_map[*i] = *i;
+  }
+
+  // the set of all canonical property names
+  std::set< std::string > prop_names;
+  for (prop_map::iterator i = keyword_map.begin();
+       i != keyword_map.end(); ++i) {
+    prop_names.insert((*i).second);
+  }
+
+  // set up DagMC's property tags based on what's been requested
+  for (std::set<std::string>::iterator i = prop_names.begin();
+       i != prop_names.end(); ++i) {
+    std::string tagname("DAGMCPROP_");
+    tagname += (*i);
+
+    Tag new_tag;
+    rval = MBI->tag_get_handle(tagname.c_str(), 0, MB_TYPE_OPAQUE, new_tag,
+                               MB_TAG_SPARSE | MB_TAG_VARLEN | MB_TAG_CREAT);
+    if (MB_SUCCESS != rval)
+      return rval;
+    property_tagmap[(*i)] = new_tag;
+  }
+
+  // now that the keywords and tags are ready, iterate over all the actual geometry groups
+  for (std::vector<EntityHandle>::iterator grp = group_handles().begin();
+       grp != group_handles().end(); ++grp) {
+
+    prop_map properties;
+    rval = parse_group_name(*grp, properties, delimiters);
+    if (rval == MB_TAG_NOT_FOUND)
+      continue;
+    else if (rval != MB_SUCCESS)
+      return rval;
+
+    Range grp_sets;
+    rval = MBI->get_entities_by_type(*grp, MBENTITYSET, grp_sets);
+    if (MB_SUCCESS != rval)
+      return rval;
+    if (grp_sets.size() == 0)
+      continue;
+
+    for (prop_map::iterator i = properties.begin();
+         i != properties.end(); ++i) {
+      std::string groupkey = (*i).first;
+      std::string groupval = (*i).second;
+
+      if (property_tagmap.find(groupkey) != property_tagmap.end()) {
+        Tag proptag = property_tagmap[groupkey];
+        const unsigned int groupsize = grp_sets.size();
+        for (unsigned int j = 0; j < groupsize; ++j) {
+          rval = append_packed_string(proptag, grp_sets[j], groupval);
```

Comment:
Should `rval` be checked here?  Otherwise it looks like `rval` is never used and is overridden for every iteration of this loop.